// node-os/apps/files/index.js
var settings = __dirname + '/../settings';
var fs = require('fs');
var _ = require('underscore');

exports.title = 'Файлы';

exports.getFiles = function (request, callback) {
	var dir = request.dir || settings.root || '/',
		files = [],
		dirs = [];
	_.each(fs.readdirSync(dir), function (filename) {
		try {
			var stat = fs.statSync(dir + filename);
			var size;
			if (stat.isDirectory()) {
				size = fs.readdirSync(dir + '/' + filename).length + ' items';
			} else {
				size = stat.size;
				if (size < 1024) {
					size = size.toFixed(2) + ' bytes';
				} else if (size >= 1024 && size < 1024 * 1024) {
					size = (size / 1024).toFixed(2) + ' kB';
				} else if (size >= 1024 * 1024 && size < 1024 * 1024 * 1024) {
					size = (size / (1024 * 1024)).toFixed(2) + ' mB';
				} else {
					size = (size / (1024 * 1024 * 1024)).toFixed(2) + ' gB';
				}
			}
			var file = {
				full: dir + filename,
				filename: filename,
				size: size,
				mtime: stat.mtime,
				isDir: stat.isDirectory(),
			};
			if (file.isDir) {
				dirs.push(file);
			} else {
				files.push(file);
			}
		} catch (err) {  }
	});
	files.sort(function (a, b) {
		if (a.filename > b.filename) {
			return 1;
		} else {
			return -1;
		}
	});
	dirs.sort(function (a, b) {
		if (a.filename > b.filename) {
			return 1;
		} else {
			return -1;
		}
	});
	files = dirs.concat(files);
	callback(files);
};
